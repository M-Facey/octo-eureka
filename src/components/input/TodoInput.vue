<script lang="ts" setup>
import TodoButton from "@/components/input/TodoButton.vue";
import IconClose from "@/components/icons/IconClose.vue";
import IconSearch from "@/components/icons/IconSearch.vue";

export interface Prop {
  label?: string;
  type: "text";
  placeholder: string;
  modelValue: string;
  isOptional?: boolean;
  hasClearButton?: boolean;
  hasSearchButton?: boolean;
  inputDataCy: string;
  clearDataCy: string;
  searchDataCy: string;
}

withDefaults(defineProps<Prop>(), {
  label: "",
  isOptional: false,
  hasClearButton: true,
  hasSearchButton: false,
});

defineEmits<{
  (e: "update:modelValue", value: string): void;
  (e: "triggerEventOnEnter"): void;
  (e: "clearTodoInput"): void;
}>();
</script>

<template>
  <div class="relative flex flex-col">
    <label v-if="label"
      >{{ label }} <span v-if="isOptional">(Optional)</span></label
    >
    <input
      :type="type"
      :placeholder="placeholder"
      :value="modelValue"
      class="w-full bg-neutral-50 dark:bg-neutral-800 border border-neutral-300 dark:border-transparent hover:border-neutral-400 dark:hover:border-neutral-500 focus:border-neutral-400 dark:focus:border-neutral-500 focus:ring-4 focus:ring-neutral-400/50 dark:focus:ring-neutral-600/50 text-neutral-800 dark:text-neutral-300 py-2 outline-none rounded-md"
      :class="{
        'px-3': !hasClearButton && !hasSearchButton,
        'pl-3 pr-10': hasClearButton || hasSearchButton,
        'pl-3 pr-20': hasClearButton && hasSearchButton,
      }"
      :data-cy="inputDataCy"
      @input="
        $emit('update:modelValue', ($event.target as HTMLInputElement).value)
      "
      @keypress.enter="$emit('triggerEventOnEnter')"
    />

    <div class="absolute top-1/2 right-2 -translate-y-1/2 flex gap-2 mb-[3px]">
      <transition name="fade">
        <todo-button
          v-if="modelValue.length !== 0 && hasSearchButton"
          button-id="search-todos"
          button-label="Search Todos"
          button-size="xs"
          :data-cy="searchDataCy"
          :is-icon-button="true"
        >
          <icon-search class="w-4" />
        </todo-button>
      </transition>
      <transition name="fade">
        <todo-button
          v-if="modelValue.length !== 0 && hasClearButton"
          button-id="clear-todo-input"
          button-label="Clear Todo Input"
          button-size="xs"
          :data-cy="clearDataCy"
          :is-icon-button="true"
          @trigger-event="$emit('update:modelValue', '')"
        >
          <icon-close class="w-4" />
        </todo-button>
      </transition>
    </div>
  </div>
</template>

<style scoped>
.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s ease;
}

.fade-enter-from,
.fade-leave-to {
  opacity: 0;
}
</style>
