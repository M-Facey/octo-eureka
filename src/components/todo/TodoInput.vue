<script lang="ts" setup>
export interface Prop {
  label?: string;
  type: "text";
  placeholder: string;
  modelValue: string;
  isOptional?: boolean;
}

withDefaults(defineProps<Prop>(), {
  label: "",
  isOptional: false,
});
defineEmits<{
  (e: "update:modelValue", value: string): void;
  (e: "triggerEvent"): void;
}>();
</script>

<template>
  <div class="flex flex-col">
    <label v-if="label"
      >{{ label }} <span v-if="isOptional">(Optional)</span></label
    >
    <input
      :type="type"
      :placeholder="placeholder"
      :value="modelValue"
      class="w-full bg-neutral-800 border border-neutral-700 hover:border-neutral-500 focus:border-neutral-500 focus:ring-4 focus:ring-neutral-600/50 text-neutral-300 px-3 py-2 outline-none rounded"
      @input="
        $emit('update:modelValue', ($event.target as HTMLInputElement).value)
      "
      @keypress.enter="$emit('triggerEvent')"
    />
  </div>
</template>
