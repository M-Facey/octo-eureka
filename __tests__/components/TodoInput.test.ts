import { mount } from "@vue/test-utils";
import { describe, it, expect } from "vitest";
import TodoInput from "@/components/input/TodoInput.vue";

describe("Todo Input Component", () => {
  it("should render", () => {
    const wrapper = mount(TodoInput, {
      props: {
        type: "text",
        placeholder: "",
        modelValue: "",
        inputDataCy: "inputCy",
        clearDataCy: "clearCy",
        searchDataCy: "searchCy",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
