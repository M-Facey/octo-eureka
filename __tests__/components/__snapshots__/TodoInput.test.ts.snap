// Vitest Snapshot v1

exports[`Todo Input Component > should render 1`] = `
"<div data-v-2f484e0f=\\"\\" class=\\"relative flex flex-col\\">
  <!--v-if--><input data-v-2f484e0f=\\"\\" type=\\"text\\" placeholder=\\"\\" class=\\"w-full h-full bg-neutral-50 dark:bg-neutral-800 border border-neutral-300 dark:border-transparent hover:border-neutral-400 dark:hover:border-neutral-500 focus:border-neutral-400 dark:focus:border-neutral-500 focus:ring-4 focus:ring-neutral-400/50 dark:focus:ring-neutral-600/50 text-neutral-800 dark:text-neutral-300 px-3 py-2 outline-none rounded-md\\" data-cy=\\"inputCy\\">
  <div data-v-2f484e0f=\\"\\" class=\\"absolute top-1/2 right-2 -translate-y-1/2 flex gap-2 mb-[3px]\\">
    <transition-stub data-v-2f484e0f=\\"\\" name=\\"fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
      <!--v-if-->
    </transition-stub>
    <transition-stub data-v-2f484e0f=\\"\\" name=\\"fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
      <!--v-if-->
    </transition-stub>
  </div>
</div>"
`;
